domain/secretbackend/state/watcher.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/changestream"
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/domain/secretbackend"
+)
+
+// WatchSecretBackendRotationChanges returns a watcher for secret backend rotation changes.
+func (s *State) WatchSecretBackendRotationChanges(wf secretbackend.WatcherFactory) (watcher.SecretBackendRotateWatcher, error) {
```

Comment:
This should be moved to the `secretbackend/service` package. When the namespace watcher triggers, we then make a call to state to load the info to pass back. ie processChanges() becomes GetBackends(backendIds...) or something.
Here's a snippet from the modelconfig service showing how a watcher is set up
```
// NewWatchableService creates a new WatchableService for interacting with
// ModelConfig and the ability to create watchers.
func NewWatchableService(
	defaultsProvider ModelDefaultsProvider,
	st State,
	watcherFactory WatcherFactory,
) *WatchableService {
	return &WatchableService{
		Service: Service{
			defaultsProvider: defaultsProvider,
			st:               st,
		},
		watcherFactory: watcherFactory,
	}
}

// Watch returns a watcher that returns keys for any changes to model
// config.
func (s *WatchableService) Watch() (watcher.StringsWatcher, error) {
	return s.watcherFactory.NewNamespaceWatcher("model_config", changestream.All, s.st.AllKeysQuery())
}
```

Reply:
The current approach can cache the sqlair statement in the watcher. 
But it's ok, I will change it.

Reply:
We can re-evaluate, but in general, if we think that it is service logic to know how to react to db changes and read subsequent data to send to a watcher client, we should keep that logic in the service layer.  The state layer should know how to read and write data given to it, but not about how that data is used. We can always have the state layer return a closure with the prepared statement which the service layer calls each time it's needed.